server/server.go

```diff
@@ -739,40 +738,85 @@ func (s *Server) globalAccount() *Account {
 	return gacc
 }
 
-// Used to setup Accounts.
-// Lock is held upon entry.
-func (s *Server) configureAccounts() error {
+// Used to setup or update Accounts.
+// Returns a map that indicates which accounts have had their stream imports
+// changed (in case of an update in configuration reload).
+// Lock is held upon entry, but will be released/reacquired in this function.
+func (s *Server) configureAccounts(reloading bool) (map[string]struct{}, error) {
+	awcsti := make(map[string]struct{})
+
 	// Create the global account.
 	if s.gacc == nil {
 		s.gacc = NewAccount(globalAccountName)
 		s.registerAccountNoLock(s.gacc)
 	}
 
-	opts := s.opts
+	opts := s.getOpts()
```

Comment:
We use lock to get, but don't copy. So when we do s.getOpts().Port lets say and its a copy of a value that seems good, but the above does not offer any protection really does it or am I missing something?

Reply:
It is to avoid possible data races, because during a config reload, s.opts gets replaced with the new version. That is, doing s.opts.Port could be reported as a data race during the reload because it would be a read of s.opts where the reload does a write of s.opts. By using getOpts(), we get under lock the pointer to the "old" options, so that prevent the data race.

Reply:
That is what I figured, but won't prevent any issues with actual values if they are being written/changed.

Reply:
It won't cause issues with actual values if they are being written/changed, but the check may be made on the "old" value. Again doing s.getOpts().Port is data-race safe, while s.opts.Port is not. For instance:
- thread 1 does `if s.opts.Port == 1234`
- thread 2 does a config reload set the new options: https://github.com/nats-io/nats-server/blob/main/server/reload.go#L907 (which result in https://github.com/nats-io/nats-server/blob/b61e411b44af43a125f89e549d0ac7337c88db89/server/server.go#L731)
This would result in the data race because of reading s.opts and writing s.opts, not because of accessing Port.

Imagine s.opts is currently a pointer to Options with address 0x123456789, during the config reload, a new separate Options object is created, say address 0x987654321. Once the reload is "complete", that's when s.opts is swapped from 0x123456789 to 0x987654321, under the s.optsMu mutex.

Reply:
So if a thread N does s.getOpts() and got pointer 0x123456789, any field in that Options object are not going to change, the updates were made in the object 0x987654321. Hence my comment of check made possibly against "old" values.